### DISCUSSION TITLE
Start/Stop FFT between copying...

### DISCUSSION THREAD

**Discussion by gcopeland** [pschatzmann/arduino-audio-tools#776] [Q&A] [answered]
Ref: https://github.com/pschatzmann/arduino-audio-tools/discussions/755

I noticed this is very close to what I'm trying to do. I'm attempting to FFT audio from a bee hive to detect swarming. But each grouping of audio requires a new window. In other words, I don't need to continuously stream the audio, but periodically sample the audio every x-minutes apart.

As I understand it, in my loop (pseudo code) I simply need to do:
loop:
  if have finished copying:
    fft.end()
    fft.begin(config)
  copier.copy()

This safely works? Should I worry of heap fragmentation?

**Comment by pschatzmann** [comment, marked as answer]
Yes that should work! 

Fragmentation might only be an issue if you allocate/deallocate memory  by some other processes after stopping and the system can't allocate the same memory again. 

However I would also expect that it might just be good enough to stop and start calling copy() maybe by using a flag. 

Also note that one copy() is not good enough! You will need to call copy() for some short period, until you get a stable result.  I guess 5 to 15 seconds might be reasonable.

**Reply by gcopeland** [reply]
This is precisely what I'm doing. I toggle a flag based on time interval (enable - sample interval) and the call of the FFT handler (disable). Which results in multiple copies per cycle (noticed four currently). However, my current implementation does not reset the window between cycles as I didn't know how. Which is what I'm trying to figure out - correctly.

Would it be possible to simply have a reset() function which resets state and src/dest pointers without have to go through the cycle of freeing and re-allocating buffers? As you appear to have at least two users with this type of need, it appears to legitimize the use case.

**Reply by pschatzmann** [reply]
If you just stop and resume, there is the chance that a small remainder from the last cycle is processed. But I would be surprised if this would disturb. 

You can also avoid this by just setting the copy size equal or a multiple of the length (converted to bytes)

If this is not good enough, feel free to submit a pull request.

**Reply by gcopeland** [reply]
Ok. Thank you for your excellent toolkit and feedback.

Cheers.

**Comment by gcopeland** [comment]
Just FYI, doing this does ultimately result in an assert failure and stack dump. It appears to trigger when there is buffer remainders. I do not believe this a viable approach. I did not spend time to investigate further. I simply reverted to the old logic, which keeps potential stale data within the window.

I may investigate further in the future. Regardless, wanted to make a note for others who may attempt this.
